app/src/main/kotlin/de/rochefort/childmonitor/ListenService.kt

```diff
@@ -0,0 +1,223 @@
+        val text = getText(R.string.listening)
+
+        // The PendingIntent to launch our activity if the user selects this notification
+        val contentIntent = PendingIntent.getActivity(this, 0,
+                Intent(this, ListenActivity::class.java), PendingIntent.FLAG_IMMUTABLE)
+
+        // Set the info for the views that show in the notification panel.
+        val b = NotificationCompat.Builder(this, CHANNEL_ID)
+        b.setSmallIcon(R.drawable.listening_notification) // the status icon
+                .setOngoing(true)
+                .setTicker(text) // the status text
+                .setContentTitle(text) // the label of the entry
+                .setContentText(name) // the contents of the entry
+                .setContentIntent(contentIntent)
+        return b.build()
+    }
+
+    private fun createNotificationChannel() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            val serviceChannel = NotificationChannel(
+                    CHANNEL_ID,
+                    "Foreground Service Channel",
+                    NotificationManager.IMPORTANCE_DEFAULT
+            )
+            notificationManager.createNotificationChannel(serviceChannel)
+        }
+    }
+
+    fun setErrorCallback(errorCallback: (() -> Unit)) {
+        this.errorCallback = errorCallback
+    }
+
+    fun setUpdateCallback(updateCallback: (() -> Unit)) {
+        this.updateCallback = updateCallback
+    }
+
+    inner class ListenBinder : Binder() {
+        val service: ListenService
+            get() = this@ListenService
+    }
+
+    private var errorCallback: (() -> Unit)? = null
+    private var updateCallback: (() -> Unit)? = null
+    private fun doListen(address: String?, port: Int) {
+        val lt = Thread {
+            Socket(address, port).use { socket ->
+                val result = runCatching { socket.getInputStream() }
+                if (result.isSuccess) {
+                    result.getOrNull()?.let {inputStream ->
+                        withAudioTrack { audioTrack ->
+                            streamAudio(inputStream, audioTrack)
+                        }
+                    }
+                }
+                else {
+                    val exception = result.exceptionOrNull()
+                    Log.e(TAG, "Failed to stream audio", exception)
+                }
+            }
+            if (!Thread.currentThread().isInterrupted) {
+                // If this thread has not been interrupted, likely something
+                // bad happened with the connection to the child device. Play
+                // an alert to notify the user that the connection has been
+                // interrupted.
+                playAlert()
+                errorCallback?.invoke()
+            }
+        }
+        this.listenThread = lt
+        lt.start()
+    }
+
+    private fun withAudioTrack(block: (AudioTrack) -> Unit): Unit {
```

Comment:
Not sure about the purpose of this method.

Also the return value is pointless (and unused).

Reply:
It sets up the audio-track for the passed block, and closes and stops the track in any case.

Reply:
I meant the purpose of extracting a method for this. 

I guess it's a matter of taste personally I don't find the code to be more readable like this.

Reply:
I am  coming from C++ / C side, and always concerned about keeping the open / close, acquire / release, or in this case startRecording / stop methods close together, and ensure that they are always called.

I was thinking, I could write also something like
```
        ... setup audioTrack ...
        Closeable { audioTrack.stop() }.use {
          ...
        }
```
but the we are like four levels of indent down, and the use is a block itself.

Reply:
> keeping the open / close, acquire / release, or in this case startRecording / stop methods close together

Okay, I can agree with that logic, but then again, such a refactoring should not change semantics _in the refactoring step_
If the semantics need to be improved this should happen in a different commit.

Reply:
I think, I've undone the refactoring.

Reply:
Unfortunately still not... The current state still suffers from the problem I outlined in [this post](https://github.com/enguerrand/child-monitor/pull/12#discussion_r1509561136)

The Socket constructor can throw an Exception which was caught in the original code and is no longer caught now.

Reply:
True, I missed that. I hope, I have handled it now in an acceptable manner.